deeppavlov/core/models/lr_scheduled_model.py

```diff
@@ -0,0 +1,345 @@
+            return self.end_val + (self.start_val - self.end_val) / 2 * cos_out
+        elif self.dec_type == DecayType.EXPONENTIAL:
+            ratio = self.end_val / self.start_val
+            return self.start_val * (ratio ** (self.iters / self.nb))
+        elif self.dec_type == DecayType.POLYNOMIAL:
+            delta_val = self.start_val - self.end_val
+            return self.end_val + delta_val * (1 - self.iters / self.nb) ** self.extra
+        elif self.dec_type == DecayType.ONECYCLE:
+            if self.iters > self.cycle_nb:
+                # decaying from end_val to start_val for cycle_nb steps
+                pct = 1 - (self.iters - self.cycle_nb) / self.cycle_nb
+                return self.start_val * (1 + pct * (self.div - 1))
+            else:
+                # raising from start_val to end_val for cycle_nb steps
+                pct = self.iters / self.cycle_nb
+                return self.start_val * (1 + pct * (self.div - 1))
+        elif self.dec_type == DecayType.TRAPEZOID:
+            if self.iters > 0.6 * self.nb:
+                # decaying from end_val to start_val for 4/10 * nb steps
+                pct = 2.5 * (self.nb - self.iters) / self.nb
+                return self.start_val * (1 + pct * (self.div - 1))
+            elif self.iters > 0.1 * self.nb:
+                # constant end_val
+                return self.end_val
+            else:
+                # raising from start_val to end_val for 1/10 * nb steps
+                pct = 10.0 * self.iters / self.nb
+                return self.start_val * (1 + pct * (self.div - 1))
+
+
+DType = Union[str, DecayType]
+
+
+class LRScheduledModel:
+    """
+    Abstract model enhanced with optimizer, learning rate and momentum
+    management and search.
+    """
+
+    @abstractmethod
+    def _init_learning_rate_variable(self):
+        pass
+
+    @abstractmethod
+    def _init_momentum_variable(self):
+        pass
+
+    @abstractmethod
+    def _update_graph_variables(self, learning_rate=None, momentum=None):
+        """
+        Update learning rate in graph if `learning_rate` is not None,
+        Update momentum in graph if `momentum` is not None
+        """
+        if learning_rate is not None:
+            # update learning rate
+            pass
+        if momentum is not None:
+            # update momentum
+            pass
+
+    def __init__(self,
+                 learning_rate: Union[None, float, Tuple[float, float]] = None,
+                 learning_rate_decay: Union[DType, Tuple[DType, Any]] = DecayType.NO,
+                 learning_rate_decay_epochs: int = 0,
+                 learning_rate_decay_batches: int = 0,
+                 learning_rate_drop_div: float = 2.0,
+                 learning_rate_drop_patience: Optional[int] = None,
+                 momentum: Union[None, float, Tuple[float, float]] = None,
+                 momentum_decay: Union[DType, Tuple[DType, Any]] = DecayType.NO,
```

Comment:
Can you add some documentation on what `Any` here could be? And maybe this `extra` can be moved to a different argument?

Reply:
@vikmary could you, please, add this info because I don't know))

Reply:
`extra` parameter is assigned as part of the `learning_rate_decay` (`momentum_decay`), because it seemed natural to assign decay type and extra together.

For example,
`learning_rate_decay` might be just `"LINEAR"`, but can be `["POLYNOMIAL", 2]` in the sense that it is polynomial with power 2.0. So extra is responsible for specifying decay type.

Do you still think it is better to move it to a separate argument?

Reply:
let's leave it now as it is